docs/orchestration.asciidoc

```diff
@@ -0,0 +1,104 @@
+
+Whenever a Pod is removed and recreated (maybe with a newer revision), the StatefulSet controller ensures PersistentVolumes that were attached to the original Pod are attached again to the replacing Pod.
+
+[id="{p}-upgrade-patterns"]
+==== Cluster upgrade patterns
+
+Depending on how the NodeSets specification are updated, ECK handles the Kubernetes resources reconciliation in various ways.
+
+* When a new NodeSet is added to the Elasticsearch resource, ECK creates the corresponding StatefulSet. It also sets up link:https://kubernetes.io/docs/concepts/configuration/secret/[Secrets] and link:https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/[ConfigMaps] to hold the Elasticsearch configuration files and TLS certificates.
+* When the node count of an existing NodeSet is increased, ECK increases the replicas of the corresponding StatefulSet.
+* When the node count of an existing NodeSet is decreased, ECK migrates data away from the corresponding Elasticsearch nodes to remove, then decreases the replicas of the corresponding StatefulSet, once data migration is over. Corresponding <<{p}-volume-claim-templates,PersistentVolumeClaims are automatically removed>>.
+* When an existing NodeSet is removed, ECK migrates data away from the corresponding Elasticsearch nodes to remove, decreases the StatefulSet replicas accordingly, then finally removes the corresponding StatefulSet.
+* When the specification of an existing NodeSet is updated (for example the Elasticsearch configuration, or the PodTemplate resources requirements), ECK performs a rolling upgrade of the corresponding Elasticsearch nodes. In order to do so, it follows link:https://www.elastic.co/guide/en/elasticsearch/reference/current/rolling-upgrades.html[Elasticsearch rolling upgrade best practices], to slowly upgrade Pods to the newest revision while preventing unavailability of the Elasticsearch cluster. In most cases, it corresponds to restarting Elasticsearch nodes one by one, and reusing the same PersistentVolume data. Note that some <<{p}-orchestration-limitations,cluster topologies may cause the cluster to be unavailable during the upgrade>>.
+* When an existing NodeSet is renamed, ECK performs the creation of a new NodeSet with the new name, and the removal of the old NodeSet, according to the NodeSet creation and removal patterns described above. Elasticsearch data is migrated away from the deprecated NodeSet before removal. The Elasticsearch resource <<{p}-update-strategy>> controls how many nodes can exist above or below the target node count during the upgrade.
+
+In any case, ECK ensures any StatefulSet operation is done in a way that is consistent with the Elasticsearch cluster orchestration best practices, by manipulating orchestration settings such as `discovery.seed_hosts`, `cluster.initial_master_nodes`, `discovery.zen.minimum_master_nodes`, `_cluster/voting_config_exclusions`.
```

Comment:
```suggestion
In every case, ECK ensures any StatefulSet operation is done in a way that is consistent with the Elasticsearch cluster orchestration best practices, by manipulating orchestration settings such as `discovery.seed_hosts`, `cluster.initial_master_nodes`, `discovery.zen.minimum_master_nodes`, and `_cluster/voting_config_exclusions` as necessary.
```

Reply:
I'm surprised we say "with Elasticsearch cluster". I would have thought it's either "with Elasticsearch" or "with the Elasticsearch cluster"?

Reply:
I think "with Elasticsearch" is fine too, but "with the Elasticsearch cluster" is incorrect in my opinion as the sentence is "consistent with <...> best practices" so it's "Elasticsearch best practices" or "Elasticsearch cluster best practices". I don't know the fancy grammar words for it but I hope it's clear what I mean.

Reply:
made it "Elasticsearch" only